---
title: "Fun Introductory Command Line Exercise: Next Generation Sequencing (NGS) Quality Analysis with Emoji"
author: "Developed by: Ray Enke, Rachael St. Jaques1, Max Maza, Caylin Murray, Sabrina Robertson, Andrew Lonsdale, & Jason Williams. Modified by BÃ¡rbara Bitarello"
output:
  html_document:
    df_print: paged
---


## Learning Goals: Use basic command line coding to:

* Introduce students to writing basic command line scripts
* Analyze & assess the quality of FASTQ formatted NGS data
* Trim/filter low quality reads in FASTQ files

## Background 

The 1st step of any Next Generation Sequencing (NGS) analysis pipeline is checking the quality of the raw sequencing reads in each FASTQ formatted file. If the sequence quality is poor, then your resulting downstream analysis will be inaccurate and misleading. 

FastQC is a popular software used to provide an overview of basic quality metrics for NGS data. In this lesson, you will use an even more universal form of communication to analyze FASTQ files, THE EMOJI ðŸ˜»ðŸ˜»ðŸ˜». 

FASTQE is a program that analyzes FASTQ files & reads out an emoji output as an indicator of the sequenceâ€™s quality in the file. So, a high quality read may look like this ðŸ˜ƒ, while this symbol ðŸ’© indicates... well you get the idea.

Like the popular FastQC software, FASTQE can be used to analyze FASTQ file quality whether itâ€™s from a genome sequencing project, an RNA-seq project, a ChIP-seq project, etc. Hereâ€™s a brief background on the in-class metagenomics project that Dr. Enkeâ€™s Bio 481 Genomics class at James Madison University is collecting data for. Garter snakes excrete sexually dimorphic pheromones to attract a mate. 

The hypothesis of their experiment is that male and female garter snakes host unique microbial communities in their musk glands that contribute to sexually dimorphic bioengineering of these pheromone molecules. We are using FASTQ files from this project but feel free to substitute your own favorite FASTQ files for this activity if you like.

The FASTQE program is limited to short read NGS data of 500bp/read or less.

## Technical requirements/limitations:

* FASTQE software is able to run natively on Mac OS or Linux computer with Anaconda installed. Since Windows does not support the use of emoticons (ðŸ˜ŸðŸ˜±ðŸ˜¿), this implementation of the lesson can be also be run in a Jupyter notebook on any machine!

* Use your own machine (linux or mac) if you have anaconda and Python version >=3 installed.

* Use jupyter lab online if you don't or prefer the online version. If so, use the "upload" button to upload the file "fastqe.ipynbd". Click on it and open it. Open also a terminal window on jupyter lab and follow the instructions.

---


## Part 1: Download FASTQ files and run fastqe

Jupyter allows you to run commands by selecting a cell and then click the play button or Cntrl+Enter.

For example, running the next cell executes the pwd (print working directory) command, which will tell you what directory this notebook is located in. Try it:

```{bash}
pwd
```


